js/application.js

```diff
@@ -126,194 +209,507 @@ $(document).ready(function(){
-        for (j=3; j>=1; j--){
+      $('#'+gridRow).children('#column'+max).attr('myAttr','cantMove');
+      while($('#'+gridRow).find("div[myAttr='canMove']").length!=0){
+        for (j=(max-1); j>=1; j--){
           gridCol="column"+j;
           gridColPrev="column"+(j+1);
-          if($('#'+gridRow).children('#'+gridCol).text()!="0"){
-            if (($('#'+gridRow).children('#'+gridColPrev).text()=='0')) {
-              sum=Number($('#'+gridRow).children('#'+gridColPrev).text())+Number($('#'+gridRow).children('#'+gridCol).text());
-              $('#'+gridRow).children('#'+gridColPrev).text(sum);
-              $('#'+gridRow).children('#'+gridCol).text(0);
-              $('#'+gridRow).children('#'+gridCol).attr('myAttr','canMove');
-              $('#'+gridRow).children('#'+gridCol).attr('combine','canCombine');
-              $('#'+gridRow).children('#'+gridColPrev).attr('combine','canCombine');
+          currentElement=$('#'+gridRow).children('#'+gridCol)
+          prevElement=$('#'+gridRow).children('#'+gridColPrev);
+          if(currentElement.text()!="0"){
+            if (prevElement.text()=='0') {
+              sum=Number(prevElement.text())+Number(currentElement.text());
+              prevElement.text(sum);
+              currentElement.text(0);
+              currentElement.attr('myAttr','canMove');
+              currentElement.attr('combine','canCombine');
+              prevElement.attr('combine','canCombine');
               var hasMoved=true;
-            } else if (($('#'+gridRow).children('#'+gridCol).text()==$('#'+gridRow).children('#'+gridColPrev).text())&&($('#'+gridRow).children('#'+gridColPrev).attr('combine')=='canCombine')&&($('#'+gridRow).children('#'+gridCol).attr('combine')=='canCombine')) {
-              sum=Number($('#'+gridRow).children('#'+gridColPrev).text())+Number($('#'+gridRow).children('#'+gridCol).text());
-              $('#'+gridRow).children('#'+gridColPrev).text(sum);
-              $('#'+gridRow).children('#'+gridCol).text(0);
+            } else if ((currentElement.text()==prevElement.text())&&(prevElement.attr('combine')=='canCombine')&&(currentElement.attr('combine')=='canCombine')) {
+              sum=Number(prevElement.text())+Number(currentElement.text());
+              prevElement.text(sum);
               totalScore += sum;
-              $('#'+gridRow).children('#'+gridCol).attr('myAttr','canMove');
-              $('#'+gridRow).children('#'+gridCol).attr('combine','canCombine');
-              $('#'+gridRow).children('#'+gridColPrev).attr('combine','cantCombine');
+              currentElement.text(0);
+              currentElement.attr('myAttr','canMove');
+              currentElement.attr('combine','canCombine');
+              prevElement.attr('combine','cantCombine');
               var hasMoved=true;
```

Comment:
Make the following a function. You are repeating this block of code for 4 times.

```js
sum=Number(prevElement.text())+Number(currentElement.text());
prevElement.text(sum);
totalScore += sum;
currentElement.text(0);
currentElement.attr('myAttr','canMove');
currentElement.attr('combine','canCombine');
prevElement.attr('combine','cantCombine');
var hasMoved=true;
```